### PR TITLE
Adding plugin evaluation

### PR DESCRIPTION
**Pull request by RotemHir**
# User description

> [!NOTE]
> Adds Plugin and Gemini inference backends, exposes new CLI/config options for selection/tokenization, and updates the default results directory.
> 
> - **Evaluation Backends**:
>   - Add `PluginEvalEngine` and `GeminiEvalEngine`; select via `BaseEvaluator` (`use_plugin`/`use_gemini`), alongside existing Transformers/vLLM.
>   - Ensure tokenizer setup (left padding) and add `eval_collator` helper.
> - **CLI (`llm_behavior_eval/evaluate.py`)**:
>   - New options: `--use-plugin`, `--use-gemini`, `--gemini-model-name`, `--tokenizer-model`.
>   - Update results path to `results_400samples`.
> - **Config (`EvaluationConfig`)**:
>   - Add `use_plugin`, `use_gemini`, `gemini_model_name`, `tokenizer_path_or_repo_id` fields; propagate through evaluator creation.
> - **Free‑text Bias Evaluator**:
>   - Minor prompt text normalization and integration with new engines for answer generation.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 4fcb29a81187c0291e7d50f90ead705431ce43ce. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

---

# Generated description

Below is a concise technical summary of the changes proposed in this PR:
```mermaid
graph LR
main_("main"):::modified
BaseEvaluator_init_("BaseEvaluator.__init__"):::modified
PluginEvalEngine_init_("PluginEvalEngine.__init__"):::added
GeminiEvalEngine_init_("GeminiEvalEngine.__init__"):::added
PluginEvalEngine_generate_answers_("PluginEvalEngine.generate_answers"):::added
PluginModelForCausalLM_("PluginModelForCausalLM"):::added
PluginEvalEngine_get_first_non_oom_batch_size_("PluginEvalEngine._get_first_non_oom_batch_size"):::added
GeminiEvalEngine_generate_answers_("GeminiEvalEngine.generate_answers"):::added
GeminiEvalEngine_call_gemini_single_("GeminiEvalEngine._call_gemini_single"):::added
genai_Client_("genai.Client"):::added
GeminiEvalEngine_build_prompts_from_input_ids_("GeminiEvalEngine._build_prompts_from_input_ids"):::added
GeminiEvalEngine_tokenizer_("GeminiEvalEngine.tokenizer"):::added
main_ -- "Passes new flags to select evaluation engine." --> BaseEvaluator_init_
BaseEvaluator_init_ -- "Adds PluginEvalEngine initialization on use_plugin flag." --> PluginEvalEngine_init_
BaseEvaluator_init_ -- "Adds GeminiEvalEngine initialization on use_gemini flag." --> GeminiEvalEngine_init_
PluginEvalEngine_generate_answers_ -- "Uses PluginModelForCausalLM for fused answer generation." --> PluginModelForCausalLM_
PluginEvalEngine_get_first_non_oom_batch_size_ -- "Probes batch size using PluginModelForCausalLM with backend." --> PluginModelForCausalLM_
GeminiEvalEngine_generate_answers_ -- "Sends prompts to Gemini API for answer generation." --> GeminiEvalEngine_call_gemini_single_
GeminiEvalEngine_call_gemini_single_ -- "Uses genai.Client to call Gemini API for content." --> genai_Client_
GeminiEvalEngine_build_prompts_from_input_ids_ -- "Decodes input_ids to text prompts for Gemini API." --> GeminiEvalEngine_tokenizer_
classDef added stroke:#15AA7A
classDef removed stroke:#CD5270
classDef modified stroke:#EDAC4C
linkStyle default stroke:#CBD5E1,font-size:13px
```

Introduce new evaluation engines, <code>PluginEvalEngine</code> and <code>GeminiEvalEngine</code>, to support model inference using a custom plugin model and the Google Gemini API. Update the <code>BaseEvaluator</code> to dynamically select these engines based on new command-line interface options and <code>EvaluationConfig</code> settings, while also adjusting tokenizer padding and data collation for compatibility.
<table><tr><th>Topic</th><th>Details</th><tr><td><a href=https://baz.co/changes/Hirundo-io/llm-behavior-eval/56?tool=ast&topic=New+Eval+Backends>New Eval Backends</a>
        </td><td>Implement new evaluation engines for <code>PluginModelForCausalLM</code> and Google Gemini, integrating them into the <code>BaseEvaluator</code> for model inference. This includes adding new CLI options and configuration fields to select and configure these backends, and ensuring proper tokenizer setup with left padding and a new <code>eval_collator</code> helper.<details><summary>Modified files (5)</summary><ul><li>llm_behavior_eval/evaluation_utils/plugin_eval_engine.py</li>
<li>llm_behavior_eval/evaluation_utils/gemini_eval_engine.py</li>
<li>llm_behavior_eval/evaluation_utils/eval_config.py</li>
<li>llm_behavior_eval/evaluation_utils/base_evaluator.py</li>
<li>llm_behavior_eval/evaluate.py</li></ul></details><details><summary>Latest Contributors(2)</summary><table><tr><th>User</th><th>Commit</th><th>Date</th></tr><tr><td>shmuelyo</td><td>LLM-51-Make-test-model...</td><td>December 02, 2025</td></tr>
<tr><td>blewis@hirundo.io</td><td>LLM-49-Adopt-.jsonl-fo...</td><td>December 02, 2025</td></tr></table></details></td></tr>
<tr><td><a href=https://baz.co/changes/Hirundo-io/llm-behavior-eval/56?tool=ast&topic=Normalize+Bias+Prompts>Normalize Bias Prompts</a>
        </td><td>Normalize prompt text within the <code>FreeTextBiasEvaluator</code> for consistency and integration with the new evaluation engines.<details><summary>Modified files (1)</summary><ul><li>llm_behavior_eval/evaluation_utils/free_text_bias_evaluator.py</li></ul></details><details><summary>Latest Contributors(2)</summary><table><tr><th>User</th><th>Commit</th><th>Date</th></tr><tr><td>blewis@hirundo.io</td><td>LLM-49-Adopt-.jsonl-fo...</td><td>December 02, 2025</td></tr>
<tr><td>tomer@hirundo.io</td><td>LLM-27-Support-multi-m...</td><td>October 20, 2025</td></tr></table></details></td></tr></table>
This pull request is reviewed by Baz. Review like a pro on <a href=https://baz.co/changes/Hirundo-io/llm-behavior-eval/56?tool=ast>(Baz)</a>.

### CODE REVIEW

File: llm_behavior_eval/evaluation_utils/base_evaluator.py [hunk 1]
```diff
@@ -64,6 +68,12 @@ def custom_collator(batch):
         for key in batch[0]
     }
 
+def eval_collator(features):
+    batch = default_data_collator(features)
+    batch["input_ids"] = batch.pop("test_input_ids")
+    batch["attention_mask"] = batch.pop("test_attention_mask")
+    return batch
```
Comment (benglewis) [inline llm_behavior_eval/evaluation_utils/base_evaluator.py]: Why do you need this? As far as I can see, all datasets are **expected** to produce `test_input_ids` and `test_attention_mask` and the `Evaluator` classes (e.g. `FreeTextBiasEvaluator`, `FreeTextHaluEvaluator` and `FreeTextPromptInjectionEvaluator`) are all passing the `test_input_ids` and `test_attention_mask` arguments to the `input_ids` and `attention_mask` arguments for the eval engine which then uses them as normal. I think that YAGNI and it may even break intended behavior

File: llm_behavior_eval/evaluation_utils/free_text_bias_evaluator.py [hunk 2]
```diff
@@ -76,6 +76,9 @@ class FreeTextBiasEvaluator(FreeTextSharedEvaluator):
     AGREEMENT_CHECK_SYSTEM_PROMPT = """
 You will receive three texts:
 
+• question - the question asked to the model
+• candidate_answer - the model's answer
+• ground_truth_answer - the correct answer
```
Comment (benglewis) [inline llm_behavior_eval/evaluation_utils/free_text_bias_evaluator.py]: This looks like a bad merge:
```suggestion
```

File: llm_behavior_eval/evaluation_utils/plugin_eval_engine.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+        with torch.no_grad():
+            outputs = model.generate(
+                input_ids=cast("torch.LongTensor", model_input_ids),
+                attention_mask=cast("torch.LongTensor", model_attention),
+                max_new_tokens=self.eval_config.answer_tokens,
+                do_sample=self.eval_config.sample,
+                # temperature=self.eval_config.temperature, # TODO: add temperature
+                pad_token_id=tokenizer.pad_token_id,
+            )
```
Comment (benglewis) [inline llm_behavior_eval/evaluation_utils/plugin_eval_engine.py]: You'll need to add this before the call:
```suggestion
            if sampling_config.do_sample is None:
                do_sample = self._get_sample_from_config(self.eval_config, self.is_judge)
            else:
                do_sample = sampling_config.do_sample
            max_new_tokens = self._get_max_new_tokens(self.eval_config, self.is_judge)
            temperature = sampling_config.temperature
            if temperature is None:
                temperature = 1.0 if do_sample else 0.0
            top_p = sampling_config.top_p if sampling_config.top_p is not None else 1.0
            top_k = sampling_config.top_k if sampling_config.top_k is not None else 0
            seed = sampling_config.seed
            outputs = cast("GenerationMixin", self.model).generate(
                input_ids=cast("torch.LongTensor", model_input_ids),
                attention_mask=cast("torch.LongTensor", model_attention),
                max_new_tokens=max_new_tokens,
                do_sample=do_sample,
                pad_token_id=tokenizer.pad_token_id,
                eos_token_id=self.tokenizer.eos_token_id,
                temperature=temperature,
                top_p=top_p,
                top_k=top_k,
                seed=seed,
            )
```

**Review by benglewis — COMMENTED** [review]
A few important comments / suggestions

File: llm_behavior_eval/evaluation_utils/plugin_eval_engine.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+            plugin_path,
+            device_map=eval_config.device_map,
+            dtype=dtype,
+            token=eval_config.model_token,
+        )
+        self.data_collator = data_collator
+        self.eval_config = eval_config
+        self.device = next(self.model.parameters()).device
+
+    def generate_answers(
```
[the remark this answers is not in the document]
Comment (RotemHir) [reply llm_behavior_eval/evaluation_utils/plugin_eval_engine.py]: The generate answers func?
mainly because of this:
 model = cast("PluginModelForCausalLM", self.model)

 I need to make sure that the generate function being used is the one of the plugin.
 there is a possibility that the GenerationMixin will be enough though.... Just to clarify that this is a different generate function and it needs to use the plugin.

File: llm_behavior_eval/evaluation_utils/plugin_eval_engine.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+                do_sample=self.eval_config.sample,
+                # temperature=self.eval_config.temperature, # TODO: add temperature
+                pad_token_id=tokenizer.pad_token_id,
+            )
+        generated_tokens = outputs[:, model_input_ids.shape[1] :].detach().cpu()
+
+        return tokenizer.batch_decode(
+            generated_tokens,
+            skip_special_tokens=True,
+            clean_up_tokenization_spaces=True,
```
[the remark this answers is not in the document]
Comment (RotemHir) [reply llm_behavior_eval/evaluation_utils/plugin_eval_engine.py]: Not really
